File: lab2.js
class Hopfield {
    constructor(images, nu = 1) {
        this.size = images[0].length;
        this.w = Array.from({ length: this.size }, () => Array(this.size).fill(0));
        this.images = images;
        this.negImages = this._getNegImages(images);
        this.nu = nu;
    }

    _getNegImages(images) {
        return images.map(image => image.map(value => -value));
    }

    train(e = 1e-6, maxIters = 10000) {
        for (let iter = 0; iter < maxIters; iter++) {
            const oldW = this.w.map(row => [...row]);

            for (const image of this.images) {
                const xT = image.map(val => [val]); // Transpose of x
                const deltaW = this._matrixMult(
                    this._scalarMult(this.nu / this.size, this._matrixSub(xT, this._matrixMult(this.w, xT))),
                    this._transpose(xT)
                );
                this.w = this._matrixAdd(this.w, deltaW);

                // Set diagonal to 0
                for (let i = 0; i < this.size; i++) {
                    this.w[i][i] = 0;
                }
            }

            // Check convergence
            const diff = this._matrixDiff(oldW, this.w);
            if (diff < e) break;
        }

        // Ensure diagonal is 0
        for (let i = 0; i < this.size; i++) {
            this.w[i][i] = 0;
        }
    }

    predict(x, maxIters = 1000) {
        const states = Array(4).fill(x.map(val => [val])); // Store states as column vectors
        let relaxationIters = 0;

        for (let iter = 0; iter < maxIters; iter++) {
            relaxationIters++;
            const newState = this._transpose(this._matrixTanh(this._matrixMult(this.w, this._transpose(states[3]))));
            states.shift();
            states.push(newState);

            if (iter >= 3) {
                const diff1 = this._matrixMaxAbs(this._matrixSub(states[0], states[2]));
                const diff2 = this._matrixMaxAbs(this._matrixSub(states[1], states[3]));

                if (diff1 < 1e-10 && diff2 < 1e-10) {
                    const imageNum = this._findImageNum(newState, this.images);
                    const negImageNum = this._findImageNum(newState, this.negImages);
                    const isNegative = negImageNum !== null;

                    return { relaxationIters, newState, imageNum: imageNum ?? negImageNum, isNegative };
                }
            }
        }

        return { relaxationIters: maxIters, newState: states[3], imageNum: null, isNegative: null };
    }

    _findImageNum(x, images) {
        return images.findIndex(image => this._arraysEqual(image, x)) ?? null;
    }

    _matrixMult(A, B) {
        const rowsA = A.length, colsA = A[0].length, colsB = B[0].length;
        const result = Array.from({ length: rowsA }, () => Array(colsB).fill(0));

        for (let i = 0; i < rowsA; i++) {
            for (let j = 0; j < colsB; j++) {
                for (let k = 0; k < colsA; k++) {
                    result[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        return result;
    }

    _scalarMult(scalar, matrix) {
        return matrix.map(row => row.map(value => value * scalar));
    }

    _matrixAdd(A, B) {
        return A.map((row, i) => row.map((val, j) => val + B[i][j]));
    }

    _matrixSub(A, B) {
        return A.map((row, i) => row.map((val, j) => val - B[i][j]));
    }

    _matrixTanh(matrix) {
        return matrix.map(row => row.map(val => Math.tanh(val)));
    }

    _matrixDiff(A, B) {
        return A.reduce((sum, row, i) => sum + row.reduce((rowSum, val, j) => rowSum + Math.abs(val - B[i][j]), 0), 0);
    }

    _matrixMaxAbs(matrix) {
        return Math.max(...matrix.flat().map(Math.abs));
    }

    _transpose(matrix) {
        return matrix[0].map((_, colIndex) => matrix.map(row => row[colIndex]));
    }

    _arraysEqual(arr1, arr2) {
        return arr1.length === arr2.length && arr1.every((val, i) => val === arr2[i]);
    }
}

function preprocessAlphabet(alphabet) {
    return alphabet.map(item => Array.isArray(item[0]) ? item.flat() : item);
}

function imageBeautifulPrint(image, rows, cols) {
    const mapped = image.map(val => (val > 0 ? '⬜' : '⬛')).join('');
    for (let i = 0; i < rows; i++) {
        console.log(mapped.slice(i * cols, (i + 1) * cols));
    }
}

// Example usage
const alphabet = preprocessAlphabet([
    [
        [-1, 1, 1, -1],
        [1, -1, -1, 1],
        [1, 1, 1, 1],
        [1, -1, -1, 1]
    ],
    [
        [1, 1, 1, -1],
        [1, -1, -1, 1],
        [1, -1, -1, 1],
        [1, 1, 1, -1]
    ],
    [
        [-1, 1, 1, -1],
        [1, -1, -1, 1],
        [1, -1, -1, 1],
        [-1, 1, 1, -1]
    ],
    [
        [-1, 1, 1, 1],
        [1, -1, -1, -1],
        [1, -1, -1, -1],
        [-1, 1, 1, 1]
    ]
]);

const network = new Hopfield(alphabet, 0.7);
network.train();

// Пример изображения с шумом для теста
const noisyImage = [1, 1, -1, -1, 1, 1, -1, 1, 1, -1, -1, 1, 1, -1, -1, 1];

// Предсказание с использованием сети
const result = network.predict(noisyImage);

// Выводим результат
console.log('Результат работы нейронной сети:');
imageBeautifulPrint(result.newState.flat(), 4, 4);

File: lab2.py
import numpy as np
# from numbers import Numbers
from pprint import pprint


def preprocess_alphabet(alphabet):
    """
    Преобразует двумерные элементы массива в одномерные.
    """
    processed_alphabet = []
    for item in alphabet:
        if isinstance(item, np.ndarray) and item.ndim == 2:
            processed_alphabet.append(item.ravel())
        else:
            processed_alphabet.append(item)
    return np.array(processed_alphabet)

class Hopfield:
    def __init__(self, images, nu: float = 1) -> None:
        self.size = images.shape[1]
        self.w = np.matrix(np.zeros((self.size, self.size)))
        self.images = images
        self.neg_images = self._get_neg_images(self.images)
        self.nu = nu

    def _get_neg_images(self, images):
        return images * (-1)      


    def train(self, e=1e-6, max_iters=10000):
        # метод дельта проекций
        for _ in range(max_iters):
            old_w = self.w.copy()       
            for image in self.images:

                x_t = np.matrix(image.copy()).T            
                self.w = self.w + (self.nu / self.size) * (x_t - (self.w @ x_t)) @ x_t.T
                np.fill_diagonal(self.w, 0)                             

            # Difference between old and new weights            
            if np.absolute(old_w - self.w).sum() < e:
                break
                                                                         
        np.fill_diagonal(self.w, 0)                                
    
    def _find_image_num(self, x, images) -> int | None:
        # x = np.sign(x)
        print(f'find image = {x}')
        mask = (images == x).all(axis=1)
        search_result = np.where(mask)[0]               
        if len(search_result) > 0:
            return search_result.item()
        return None

    def predict(self, x, max_iters: int = 1000):
        states = [np.matrix(x.copy())] * 4        
        relaxation_iters = 0

        for _iter in range(max_iters):
            relaxation_iters += 1            
            # new_state = np.sign(np.tanh(self.w @ states[-1].T)).T
            new_state = np.tanh(self.w @ states[-1].T).T
            # new_state = np.where(new_state > 0, 1, -1)
            # убрать знаковую
    
            states.append(new_state)
            states.pop(0)

            if _iter < 3:
                continue

            
            # if (np.mean(np.absolute(states[0]- states[2])) < 1e-8
            #     and np.mean(np.absolute(states[1]- states[3])) < 1e-8):
            
            # np.absolute(states[0]- states[2]) и берешь максимум и справниваешь с ошибкой
            
            # np.array_equal(states[0], states[2]) and np.array_equal(states[1], states[3])
            if np.absolute(states[0] - states[2]).max() < 1e-10 and np.absolute(states[1]- states[3]).max() < 1e-10:
                # print(new_state)
                image_num = self._find_image_num(new_state, self.images)
                neg_image_num = self._find_image_num(new_state, self.neg_images)
                is_negative = neg_image_num is not None

                return (relaxation_iters, new_state,
                        (image_num if image_num is not None else neg_image_num),
                        is_negative)

        # Если максимум итераций достигнут
        return max_iters, new_state, None, None

def is_negative(image, neg_image) -> bool:
    return(np.array_equal(image, neg_image))


alphabet = np.array(
    [
        [
            [-1, 1, 1, -1],
            [1, -1, -1, 1],
            [1, 1, 1, 1],
            [1, -1, -1, 1]
        ],
        [
            [1, 1, 1, -1],
            [1, -1, -1, 1],
            [1, -1, -1, 1],
            [1, 1, 1, -1]
        ],
        [
            [-1, 1, 1, -1],
            [1, -1, -1, 1],
            [1, -1, -1, 1],
            [-1, 1, 1, -1]
        ],
        [
            [-1, 1, 1, 1],
            [1, -1, -1, -1],
            [1, -1, -1, -1],
            [-1, 1, 1, 1]
        ]
    ]
)
alphabet = preprocess_alphabet(alphabet)


network = Hopfield(alphabet, 0.7)
network.train()

def image_beautiful_print(image, rows, cols):
    image = np.sign(image)
    image = image.astype(np.object_)
    image[image == 1] = '⬜'
    image[image == -1] = '⬛'
    image = image.reshape(rows, cols)
    image_list = image.tolist()
    for row in image_list:
        print(''.join(row))


# test_image = np.array([1, -1, 1, -1, 1, 1, -1, 1, 1, 1, 1, 1, 1, -1, -1, 1])
test_image = np.array([-1, -1, 1, 1, -1, -1, -1, 1, 1, 1, 1, 1, 1, -1, -1, 1])

r_iters, state, image_idx, is_negative = network.predict(test_image, 10_000)
if image_idx:
    predicted_img = network.neg_images[image_idx] if is_negative else network.images[image_idx]
else:
    predicted_img = state
print(f'Prediction success? - {True if image_idx is not None else False}')
print(f'Image number (index begins from 0): {image_idx}.\n')
print(f'Image: {predicted_img}.\nIters for relax: {r_iters}')
print(f'Is this image a negative? - {is_negative}')
print('ORIGINAL IMAGE')
image_beautiful_print(test_image, 4, 4)
print('PREDICTED IMAGE')
image_beautiful_print(predicted_img, 4, 4)
print('NEGATIVE OF PREDICTED IMAGE')
image_beautiful_print(predicted_img * (-1), 4, 4)